loadfile/loadfile.go

```diff
@@ -7,41 +7,168 @@ import (
+type ContextFile struct {
+	ID           string
+	AbsolutePath string
+	Content      []byte
+}
+
+type fileCache struct {
+	rootDir string
+	files   map[string]ContextFile
 }
 
-// AbsPathsWithContext creates a map of absolute filepaths. If a required
-// file is not provided with an absolute path, then it is joined with the
-// root directory.
-func AbsPathsWithContext(rootDir string, requiredFiles map[string]string) (map[string]string, error) {
+// FileCache is a container of ContextFiles, and a context (root)
+// directory path to be used in conjunction with files that do not
+// provide an absolute file path.
+type FileCache interface {
+	RootDir() string
+	LoadContext() error
+	GetByKey(fileKey string) (*ContextFile, bool)
+	AbsPathByKey(fileKey string) *string
+	ContentByKey(fileKey string) []byte
+	Contents() map[string][]byte
+	AbsPaths() map[string]string
+	Files() map[string]ContextFile
+}
+
+// NewFileCacheUsingContext creates a mapping of files to their absolute paths and
+// file contents. rootDir is the context directory (root directory) in
+// which the files can be found. 'files' is a mapping of the desired file
+// key to a relative or absolute file path.
+func NewFileCacheUsingContext(rootDir string, files map[string]string) (FileCache, error) {
 	absDir, err := filepath.Abs(rootDir)
 	if err != nil {
 		return nil, fmt.Errorf("error determining context directory absolute path %s (%w)", rootDir, err)
 	}
-	requiredFilesAbs := map[string]string{}
-	for key, f := range requiredFiles {
+	filesAbsPaths := map[string]ContextFile{}
+	for key, f := range files {
 		abspath := f
 		if !filepath.IsAbs(f) {
 			abspath = filepath.Join(absDir, f)
 		}
-		requiredFilesAbs[key] = abspath
+		filesAbsPaths[key] = ContextFile{
+			ID:           f,
+			AbsolutePath: abspath,
+		}
+	}
+	return &fileCache{
+		rootDir: absDir, files: filesAbsPaths}, nil
+}
+
+// NewFileCache creates a file cache where the fileContents is
+// a mapping of paths to file contents. The keys in fileContents
+// become the keys in Files, and the absolute paths and ID for
+// each context file.
+func NewFileCache(rootDir string, fileContents map[string][]byte) FileCache {
+	files := map[string]ContextFile{}
+	for key, content := range fileContents {
+		files[key] = ContextFile{
+			ID:           key,
+			AbsolutePath: key,
+			Content:      content,
+		}
+	}
+	return &fileCache{
+		rootDir: rootDir,
+		files:   files,
+	}
+}
+
+// LoadContext reads the content of each context file into Content.
+func (fc *fileCache) LoadContext() error {
+	result := map[string]ContextFile{}
+	var err error
+	for key, cf := range fc.files {
+		absPath := cf.AbsolutePath
+		fileData, err := os.ReadFile(filepath.Clean(absPath))
+		if err != nil {
+			return fmt.Errorf("error reading file %s (%w)", absPath, err)
+		}
+		result[key] = ContextFile{
+			ID:           cf.ID,
+			AbsolutePath: cf.AbsolutePath,
+			Content:      fileData,
+		}
+	}
+	fc.files = result
```

Comment:
Why do you construct a new value for `fc.files` instead of just updating each `cf.Content`?

Reply:
That broke the code. For some reason, when you leave the scope of that function, the `Content` for each of its files would still be `nil`. Maybe `fileCache.files` would need to be a pointer to `map[string]ContextFile` for that approach to work. I'm not 100% sure.

Reply:
That's weird.  😢